components/sync_manager/src/manager.rs

```diff
@@ -0,0 +1,193 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::*;
+use crate::msg_types::{ServiceStatus, SyncParams, SyncResult};
+use logins::PasswordEngine;
+use places::{bookmark_sync::store::BookmarksStore, history_sync::store::HistoryStore, PlacesApi};
+use std::collections::HashMap;
+use std::sync::Mutex;
+use std::sync::{atomic::AtomicUsize, Arc, Weak};
+use sync15::MemoryCachedState;
+
+const LOGINS_ENGINE: &str = "passwords";
+const HISTORY_ENGINE: &str = "history";
+const BOOKMARKS_ENGINE: &str = "bookmarks";
+
+pub struct SyncManager {
+    mem_cached_state: Option<MemoryCachedState>,
+    places: Weak<PlacesApi>,
+    logins: Weak<Mutex<PasswordEngine>>,
```

Comment:
The fact logins has a Mutex but PlacesApi doesn't seems like there's some inconsistency here that should be abstracted away lower?

(The fact we can't just work with traits surprises me too, but I guess there's a good reason hidden away somewhere)